staging/src/k8s.io/apimachinery/pkg/util/strategicpatch/patch_test.go

```diff
@@ -654,6 +654,86 @@ func TestCustomStrategicMergePatch(t *testing.T) {
 	}
 }
 
+// TestStrategicMergeTreatNilCollectionAsEmpty tests the behavior where nil and empty collections in
+// the Go struct are treated equivalently.
+func TestStrategicMergeTreatNilCollectionAsEmpty(t *testing.T) {
+	for i, tc := range []struct {
+		original *MergeItem
+		modified *MergeItem
+		current  *MergeItem
+		want     []byte
+	}{{
+		// original->modified deletes the last key, but current has a custom key.
+		original: &MergeItem{SimpleMap: map[string]string{"original-key": "1"}},
+		modified: &MergeItem{SimpleMap: map[string]string{}},
+		current:  &MergeItem{SimpleMap: map[string]string{"original-key": "1", "custom-key": "2"}},
+		want:     []byte(`{"SimpleMap":{"original-key":null}}`),
+	}, {
+		// original->modified deletes the last key (with simulated omitempty behavior), but current has
+		// a custom key.
+		original: &MergeItem{SimpleMap: map[string]string{"original-key": "1"}},
+		modified: &MergeItem{}, // SimpleMap is nil (simulated omitempty).
+		current:  &MergeItem{SimpleMap: map[string]string{"original-key": "1", "custom-key": "2"}},
+		want:     []byte(`{"SimpleMap":{"original-key":null}}`),
+	}, {
+		// original->modified deletes the last key (with simulated omitempty behavior).
+		original: &MergeItem{SimpleMap: map[string]string{"original-key": "1"}},
+		modified: &MergeItem{}, // SimpleMap is nil (simulated omitempty).
+		current:  &MergeItem{SimpleMap: map[string]string{"original-key": "1"}},
+		want:     []byte(`{"SimpleMap":{"original-key":null}}`), // Note: map could also be completely removed.
+	}, {
+		// original->modified and original->current delete the last key, but modified adds a new key.
+		original: &MergeItem{SimpleMap: map[string]string{"original-key": "1"}},
+		modified: &MergeItem{SimpleMap: map[string]string{"custom-key": "2"}},
+		current:  &MergeItem{SimpleMap: map[string]string{}},
+		want:     []byte(`{"SimpleMap":{"custom-key":"2","original-key":null}}`),
+	}, {
+		// original->modified and original->current delete the last key (with simulated omitempty
+		// behavior), but modified adds a new key.
+		original: &MergeItem{SimpleMap: map[string]string{"original-key": "1"}},
+		modified: &MergeItem{SimpleMap: map[string]string{"custom-key": "2"}},
+		current:  &MergeItem{}, // SimpleMap is nil (simulated omitempty).
+		want:     []byte(`{"SimpleMap":{"custom-key":"2","original-key":null}}`),
+	}, {
+		// original->modified deletes the last key from patchMergeKey slice, but current has a custom
+		// key.
+		original: &MergeItem{MergingList: []MergeItem{{Name: "original-key"}}},
+		modified: &MergeItem{MergingList: []MergeItem{}},
+		current:  &MergeItem{MergingList: []MergeItem{{Name: "original-key"}, {Name: "custom-key"}}},
+		want:     []byte(`{"MergingList":[{"$patch":"delete","name":"original-key"}]}`),
```

Comment:
The same result here. It's the next test case that would yield a different result (actually a conflict). I agree in principle that it's a conflict if you do want to explicitly delete the map. But due to `omitempty` directives sprinkled throughout the api objects, in practice you will often end up with nil maps and lists when working with deserialized API objects.